packages/shopify-app-remix/src/server/authenticate/admin/helpers/handle-embedded-client-error.ts

```diff
@@ -0,0 +1,65 @@
+import {HttpResponseError} from '@shopify/shopify-api';
+import {redirect} from '@remix-run/server-runtime';
+
+import type {HandleAdminClientError} from '../../../clients/admin/types';
+
+// import {redirectToAuthPage} from './redirect-to-auth-page';
+
+interface HandleClientErrorOptions {
+  request: Request;
+}
+
+export function handleEmbeddedClientErrorFactory({
+  request,
+}: HandleClientErrorOptions): HandleAdminClientError {
+  return async function handleClientError({
+    error,
+    params,
+    session,
+  }): Promise<never> {
+    if (error instanceof HttpResponseError !== true) {
+      params.logger.debug(
+        `Got a response error from the API: ${error.message}`,
+      );
+      throw error;
+    }
+
+    params.logger.debug(
+      `Got an HTTP response error from the API: ${error.message}`,
+      {
+        code: error.response.code,
+        statusText: error.response.statusText,
+        body: JSON.stringify(error.response.body),
+      },
+    );
+
+    console.log('error.response', JSON.stringify(error.response));
+
+    if (error.response.code === 401) {
```

Comment:
When making Admin API requests, a `401` signifies that the access token is invalid. In this scenario, we want to perform token exchange. So we delete the access token and return a `401` with `X-Shopify-Invalid-Session` so that the augmented fetch can retry the request, consequently triggering token exchange.